proof/invariant-abstract/IpcDet_AI.thy

```diff
@@ -87,111 +87,68 @@ lemma sfi_tcb_at [wp]:
     \<lbrace>\<lambda>_. tcb_at t\<rbrace>"
   by (wpsimp simp: send_fault_ipc_def)
 
-lemma sort_queue_valid_ep_helper:
-  "distinct list \<Longrightarrow> (a, b) \<in> set (zip list' list) \<Longrightarrow> (a', b) \<in> set (zip list' list) \<Longrightarrow> a = a'"
-  apply (induct list arbitrary: list')
-   apply clarsimp
-  apply (clarsimp simp: zip_Cons)
-  apply (erule disjE, fastforce dest!: in_set_zipE)
-  apply (erule disjE, fastforce dest!: in_set_zipE, clarsimp)
+lemma tcb_ep_find_index_wp:
+  "\<lbrace>\<lambda>s. (\<forall>i j. 0 \<le> i \<and> i \<le> Suc sz \<longrightarrow>
+               (\<forall>tcb tcba. ko_at (TCB tcb) tptr s \<and> ko_at (TCB tcba) (queue ! j) s \<longrightarrow>
+                           (0 \<le> j \<and> j < i \<longrightarrow> tcb_priority tcba \<ge> tcb_priority tcb) \<longrightarrow>
+                           (i < j \<and> j \<le> sz \<longrightarrow> tcb_priority tcba < tcb_priority tcb) \<longrightarrow> Q i s))\<rbrace>
+   tcb_ep_find_index tptr queue sz \<lbrace>Q\<rbrace>"
```

Comment:
I'm a bit confused, doesn't this lemma only make sense if you know that the queue was previously sorted? I thought that was the invariant we were trying to avoid?